File: executable-operator/executable-basic/src/main/java/fdu/daslab/service/client/SchedulerServiceClient.java
package fdu.daslab.service.client;

import fdu.daslab.thrift.master.SchedulerService;
import org.apache.thrift.TException;
import org.apache.thrift.protocol.TBinaryProtocol;
import org.apache.thrift.protocol.TMultiplexedProtocol;
import org.apache.thrift.protocol.TProtocol;
import org.apache.thrift.transport.TSocket;
import org.apache.thrift.transport.TTransport;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.util.HashMap;

/**
 * 调用master所需要的client
 *
 * @author 唐志伟
 * @version 1.0
 * @since 2020/9/23 3:00 PM
 */
public class SchedulerServiceClient {

    private Logger logger = LoggerFactory.getLogger(SchedulerServiceClient.class);

    // 调用的client
    private SchedulerService.Client client;
    private TTransport transport;
    // 当前client的ID标识
    private String stageId;

    // 标记是否是debug模式，不尝试和master连接
    private boolean isDebug = false;

    public SchedulerServiceClient(String stageId, String masterHost, Integer masterPort) {
        // 如果参数为空，推断出是本地模式，不会和master建立thrift连接
        if (stageId == null && masterHost == null
                && masterPort == null) {
            isDebug = true;
        } else {
            this.transport = new TSocket(masterHost, masterPort);
            TProtocol protocol = new TBinaryProtocol(transport);
            // 由于多个服务绑定一个端口，需要使用multiplex的协议
            TMultiplexedProtocol tMultiplexedProtocol = new TMultiplexedProtocol(protocol, "SchedulerService");
            this.client = new SchedulerService.Client(tMultiplexedProtocol);
            this.stageId = stageId;
        }
    }

    public boolean checkDebug(String info) {
        if (isDebug) {
            logger.info("debug mode: " + info);
        }
        return !isDebug;
    }

    // 暂时没有数据需要上报
    public void postStarted() throws TException {
        if (checkDebug("postStarted")) {
            transport.open();
            client.postStageStarted(this.stageId, new HashMap<>());
            transport.close();
        }
    }

    // 暂时没有数据需要上报
    public void postCompleted() throws TException {
        if (checkDebug("postCompleted")) {
            transport.open();
            client.postStageCompleted(this.stageId, new HashMap<>());
            transport.close();
        }
    }

    // 数据准备完，上传数据的位置信息
    public void postDataPrepared() throws TException {
        if (checkDebug("postDataPrepared")) {
            transport.open();
            client.postDataPrepared(this.stageId);
            transport.close();
        }
    }
}

File: executable-operator/executable-basic-python/main/service/client/SchedulerServiceClient.py
# import logging
# from executable.basic.utils.Logger import Logger
# from thrift.transport.TSocket import TSocket
# from thrift.protocol.TBinaryProtocol import TBinaryProtocol
# from thrift.protocol.TMultiplexedProtocol import TMultiplexedProtocol
# from thriftGen.master.SchedulerService import Client
#
#
# class SchedulerServiceClient(object):
#     """
#     Description:
#         实现thrift的Client类
#     Attributes:
#         1. stageId: 必须要在服务器的redis中注册过才能执行方法调用
#         2. masterHost: clic-master的地址
#         3. masterPort: clic-master的端口
#     """
#     def __init__(self, stageId, masterHost, masterPort):
#         if (stageId is None) and (masterPort is None) and (masterHost is None):
#             self.isDebug = True
#         else:
#             self.isDebug = False
#             self.transport = TSocket(host=masterHost, port=masterPort)
#             protocol = TBinaryProtocol(self.transport)
#             tMultiplexedProtocol = TMultiplexedProtocol(protocol, "SchedulerService")
#             self.client = Client(tMultiplexedProtocol)
#             self.stageId = stageId
#             self.logger = Logger('ServiceClientLogger', logging.DEBUG).logger
#
#     def checkDebug(self, info):
#         if self.isDebug:
#             self.logger.debug("debug mode: {}".format(info))
#         return not self.isDebug
#
#     def postStarted(self):
#         if self.checkDebug("postStarted"):
#             self.transport.open()
#             self.client.postStageStarted(self.stageId, {})
#             self.transport.close()
#
#     def postCompleted(self):
#         if self.checkDebug("postCompleted"):
#             self.transport.open()
#             self.client.postStageCompleted(self.stageId, {})
#             self.transport.close()
#
#     def postDataPrepared(self):
#         if self.checkDebug("postDataPrepared"):
#             self.transport.open()
#             self.client.postDataPrepared(self.stageId)
#             self.transport.close()
